### ISSUE TITLE
[BUG] long daemon_id could make ipc_socket exceed unix socket filename length limitation (man 7 unix) 

### ISSUE THREAD

**Issue by Cj-bc** [elkowar/eww#387]
### Checklist before submitting an issue

- [X] I have searched through the existing [closed and open issues](https://github.com/elkowar/eww/issues?q=is%3Aissue) for eww and made sure this is not a duplicate
- [X] I have specifically verified that this bug is not a common [user error](https://github.com/elkowar/eww/issues?q=is%3Aissue+label%3Ano-actual-bug+is%3Aclosed)
- [X] I am providing as much relevant information as I am able to in this bug report (Minimal config to reproduce the issue for example, if applicable)

### Description of the bug

My `~/.config/eww` is symbolic link to 
`~/Documents/ghq/github.com/Cj-bc/dotfiles/dotfiles/submodules/dotfiles-2.0/config/eww-1920`.

When I run `eww daemon`, the command outputs as normal, but actually
socket file isn't created.

After investigating for a while by print-debugging,
It seems like if the config directory is symbolic link to somewhere, `ipc_socket_file`
can't be created. (I've confirmed this problem with "reproducing issue")
Note that log file is still created.

### Reproducing the issue

1. Get eww source code from this repository
2. Do `cargo build` to build it.
3. Confirm that `./target/debug/eww daemon -c example/eww-bar` generates ipc socket correctly
4. `ln -s $(pwd)/example/eww-bar example/testing` (Name does not matter, I choose same length name)
5. Try `./target/debug/eww daemon -c example/testing` and check if ipc socket file exists. (Which in this case, it doesn't)

### Expected behaviour

ipc socket is created as expected, and daemon works fine.

### Related issues

#243 looks like similar issue, which states that we can't use symbolic linked
config directory.

### My previous thought: Filename length limitation
When I modify `crates/eww/src/main.rs` to use constant path for `ipc_socket_file`, it worked:

```diff
diff --git a/crates/eww/src/main.rs b/crates/eww/src/main.rs
index bbe9a38..f8f678e 100644
--- a/crates/eww/src/main.rs
+++ b/crates/eww/src/main.rs
@@ -221,7 +221,7 @@ impl EwwPaths {
	     ipc_socket_file: std::env::var("XDG_RUNTIME_DIR")
		 .map(std::path::PathBuf::from)
		 .unwrap_or_else(|_| std::path::PathBuf::from("/tmp"))
-                .join(format!("eww-server_{}", daemon_id)),
+               .join("eww-server_test"),
	 })
     }
```

[My commit on fork](https://github.com/Cj-bc/eww/commit/de2ebce75adc5b528fda2ade637962dea760059b)

Once before, I've thought if filename of socket file exceeds filename-length limitation of filesystem,
but as log file is generated correctly, I'm now thinking it was not the case.

### environment information

```sh
$ rustup show
Default host: x86_64-unknown-linux-gnu
rustup home:  /home/me/.rustup

installed toolchains
--------------------

nightly-2021-08-10-x86_64-unknown-linux-gnu
nightly-x86_64-unknown-linux-gnu (default)

active toolchain
----------------

nightly-x86_64-unknown-linux-gnu (default)
rustc 1.53.0-nightly (b84932674 2021-04-21)


```

- X Server
- Archlinux

**Comment by viandoxdev** [comment]
That's weird, I've had my `~/.config/eww` symlinked for a while without any problems. I can't seem to reproduce your issue either. 

I don't know if this really has to do with symlinks though, because the only link between the two is the filename, which contains the base64 encoded absolute path to the linked directory. Are you sure the socket file isn't being created somewhere else ? (in `/tmp` for example). This could be caused by wrong permissions making `XDG_RUNTIME_DIR` unwritable for some reason, but if that were the case eww probably wouldn't be the only broken thing. What kind of errors appear when using eww with a symlinked config  ?

**Comment by druskus20** [comment]
I also have my eww config symlinked with stow. Shouldn't be a problem as far as I know

**Comment by Cj-bc** [comment]
Sorry, I tested again with the same commit(8336bd0) and now it doesn't reproduce anymore...
I've confirmed that I'm using that specific commit by `git show HEAD` and `git status --short`.
Even worse, It no longer create socket even if I use `example/eww-bar` directly as config directory.
It seems like it is my environment issue, but I'm trying to dig into it.

Let me know If there's something that I can try or needed information as I don't have much experience on Rust.

And also, please let me know if something is not clear, my English skill may not be enough.

---

> Are you sure the socket file isn't being created somewhere else ? (in /tmp for example)

I didn't think about that case... I only checked the path eww told me in its output of `ewww daemon` command (`ipc-socket` in the output below)

```
$ ./target/debug/eww --debug daemon --config examples/eww-bar
 2022-02-04T09:13:06.185Z INFO  eww > Initializing Eww server. (/run/user/1000/eww-server_L2hvbWUvdGhlX2NhdC9Eb2N1bWVudHMvZ2hxL2dpdGh1Yi5jb20vZWxrb3dhci9ld3cvZXhhbXBsZXMvZXd3LWJhcg==)
Run `./target/debug/eww logs` to see any errors while editing your configuration.
 2022-02-04T09:13:06.185Z INFO  eww::server > Loading paths: config-dir: /home/me/Documents/ghq/github.com/elkowar/eww/examples/eww-bar, ipc-socket: /run/user/1000/eww-server_L2hvbWUvdGhlX2NhdC9Eb2N1bWVudHMvZ2hxL2dpdGh1Yi5jb20vZWxrb3dhci9ld3cvZXhhbXBsZXMvZXd3LWJhcg==, log-file: /home/me/.cache/eww_L2hvbWUvdGhlX2NhdC9Eb2N1bWVudHMvZ2hxL2dpdGh1Yi5jb20vZWxrb3dhci9ld3cvZXhhbXBsZXMvZXd3LWJhcg==.log

```

But I think it isn't permission problem, because modified version of eww could make socket there:

```
$ # eww command below is modified version
$ eww --debug daemon
 2022-02-04T09:26:50.973Z INFO  eww > Initializing Eww server. (/run/user/1000/eww-server_test)
Run `eww logs` to see any errors while editing your configuration.
 2022-02-04T09:26:50.973Z INFO  eww::server > Loading paths: config-dir: /home/me/Documents/ghq/github.com/Cj-bc/dotfiles/dotfiles/submodules/dotfiles-2.0/config/eww-1920, ipc-socket: /run/user/1000/eww-server_test, log-file: /home/me/.cache/eww_test.log
$ file /run/user/1000/eww-server_test
/run/user/1000/eww-server_test: socket
```

>  What kind of errors appear when using eww with a symlinked config ?

When I run `eww daemon`, it doesn't output any error message.
When I tried to use commands such as `eww kill`, it told me the error below:

```
$ ./target/debug/eww --config examples/eww-bar kill
2022-02-04T09:04:33.508Z ERROR eww::error_handling_ctx > Failed to connect to daemon
```

---

I've tried to trace where the program stops by adding `log::debug!`/`log::info!` in some place.
In `crates/eww/src/ipc_server.rs`, I added log like this:

```diff
  pub async fn run_server<P: AsRef<std::path::Path>>(evt_send: UnboundedSender<app::DaemonCommand>, socket_path: P) -> Result<()> {
+     log::info!("[Cj-bc] start of ipc_server::run_server");
      let socket_path = socket_path.as_ref();
+     log::info!("[Cj-bc] before UnixListener::bind()");
      let listener = { tokio::net::UnixListener::bind(socket_path)? };
      log::info!("IPC server initialized");
      crate::loop_select_exiting! {
```

And After running it by `cargo build && ./target/debug/eww -c ./example/eww-bar`, I read and watched the log for a while.


```
$ time tail -f -n 6 /home/me/.cache/eww_L2hvbWUvdGhlX2NhdC9Eb2N1bWVudHMvZ2hxL2dpdGh1Yi5jb20vZWxrb3dhci9ld3cvZXhhbXBsZXMvZXd3LWJhcg==.log
┏━━━━━━━━━━━━━━━━━━━━━━━┓
┃Initializing eww daemon┃
┗━━━━━━━━━━━━━━━━━━━━━━━┛
    
 2022-02-05T09:28:54.178Z INFO  eww::ipc_server > [Cj-bc] start of ipc_server::run_server
 2022-02-05T09:28:54.179Z INFO  eww::ipc_server > [Cj-bc] before UnixListener::bind()
^C

real	39m2.624s
user	0m0.005s
sys	0m0.001s
```

I waited almost 40 minutes, but it didn't print out `"IPC server initialized"` log.
So I think some problem have occured while binding the socket.

**Comment by viandoxdev** [comment]
Have you looked at this ?
![image](https://i.imgur.com/dDKw1uR.png)
(man 7 unix)

For some obscure reason unix socket files names are limited to 108 characters, if that limit where to be hit, it would:
 - not create the socket path, because it can't
 - still create the log file, as it isn't a socket and is limited to 255 characters by the filesystem
 - work when replaced by an hard coded, shorter name (i.e. `eww-server_test`)
 Which seems to be exactly your case.

If that is problem, then this is definitely an eww bug, which could be fixed by limiting the `daemon_id` () string length to something like 80 characters.

**Comment by Cj-bc** [comment]
> For some obsure reason unix socket files names are limited to 108 characters.

ahh I didn't realize this!
I've checked filesystem's limitation, but I never thought socket file have different limitation.
I'll try this way and report result here.

**Comment by Cj-bc** [comment]
I tried some value and confirmed that in my environment `daemon_id` should be less than 81 bytes so that absolute path (which will be `/run/user/1000/eww-server_{daemon_id}`) will be 108 bytes or less.
If I make it 82 bytes the problem wasn't solved.
So I think that socket filename length limitation was the problem.

```diff
diff --git a/crates/eww/src/main.rs b/crates/eww/src/main.rs
index bbe9a38..0d5860d 100644
--- a/crates/eww/src/main.rs
+++ b/crates/eww/src/main.rs
@@ -210,7 +210,8 @@ impl EwwPaths {
         }
 
         let config_dir = config_dir.canonicalize()?;
-        let daemon_id = base64::encode(format!("{}", config_dir.display()));
+        let mut daemon_id = String::from("abcdefghijklmnopqrstyvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"); // For test, ensure that this is long enough
+        daemon_id.truncate(81); // if I change 81 to 82, it won't work
 
         Ok(EwwPaths {
             config_dir,
```

_But_ I notice that even if we shorten `daemon_id` to meet the limitation, if `XDG_RUNTIME_DIR`(which is in my case `/run/user/1000`) is too long it won't work again.
I don't know how many people actually modify their `XDG_RUNTIME_DIR`, so I'm not sure whether we have to care about this case.

**Comment by viandoxdev** [comment]
`XDG_RUNTIME_DIR` should very rarely be anything else, its automatically set most of the time, but a warning never hurts.
I made a patch (I'll probably make a PR soon) that should do it:

```diff
diff --git a/crates/eww/src/main.rs b/crates/eww/src/main.rs
index f1290c3..1c58d15 100644
--- a/crates/eww/src/main.rs
+++ b/crates/eww/src/main.rs
@@ -17,6 +17,8 @@ use anyhow::{bail, Context, Result};
 use daemon_response::{DaemonResponse, DaemonResponseReceiver};
 use opts::ActionWithServer;
 use std::{
+    collections::hash_map::DefaultHasher,
+    hash::{Hash, Hasher},
     os::unix::net,
     path::{Path, PathBuf},
     time::Duration,
@@ -210,7 +212,22 @@ impl EwwPaths {
         }
 
         let config_dir = config_dir.canonicalize()?;
-        let daemon_id = base64::encode(format!("{}", config_dir.display()));
+
+        let mut hasher = DefaultHasher::new();
+        format!("{}", config_dir.display()).hash(&mut hasher);
+        // daemon_id is a hash of the config dir path to ensure that, given a normal XDG_RUNTIME_DIR,
+        // the absolute path to the socket stays under the 108 bytes limit. (see #387, man 7 unix)
+        let daemon_id = format!("{:x}", hasher.finish());
+
+        let ipc_socket_file = std::env::var("XDG_RUNTIME_DIR")
+            .map(std::path::PathBuf::from)
+            .unwrap_or_else(|_| std::path::PathBuf::from("/tmp"))
+            .join(format!("eww-server_{}", daemon_id));
+
+        // 100 as the limit isn't quite 108 everywhere (i.e 104 on BSD or mac)
+        if format!("{}", ipc_socket_file.display()).len() > 100 {
+            log::warn!("The IPC socket file's absolute path exceeds 100 bytes, the socket may fail to create.");
+        }
 
         Ok(EwwPaths {
             config_dir,
@@ -218,10 +235,7 @@ impl EwwPaths {
                 .map(PathBuf::from)
                 .unwrap_or_else(|_| PathBuf::from(std::env::var("HOME").unwrap()).join(".cache"))
                 .join(format!("eww_{}.log", daemon_id)),
-            ipc_socket_file: std::env::var("XDG_RUNTIME_DIR")
-                .map(std::path::PathBuf::from)
-                .unwrap_or_else(|_| std::path::PathBuf::from("/tmp"))
-                .join(format!("eww-server_{}", daemon_id)),
+            ipc_socket_file,
         })
     }
 

```

**Comment by Cj-bc** [comment]
@viandoxdev thanks for your work!
As I'm not a Rust programmer, it helps me a lot

**Comment by elkowar** [comment]
Should be fixed by #396

**Comment by Cj-bc** [comment]
I tried current master(fb0e57a) that contains #396 and confirmed that it now works.
Thank you a lot, @elkowar and @viandoxdev !!
